### DISCUSSION TITLE
Merge in Snyder metadata to Reece cvrs

### DISCUSSION THREAD

**Discussion by kuriwaki** [kuriwaki/cvr_harvard-mit_scripts#137] [General]
I will take the lead on proposing/drafting something for this. It probably should be a parquet-to-parquet join instead of a dataframe-to-dataframe join. And ideally we do it before the parquet goes into `CVR_parquet`

**Comment by mreece13** [comment]
Waiting on data from @kuriwaki then I will work on addressing this

**Comment by kuriwaki** [comment]
@mreece13 I am thinking about this, but one issue is that our metadata does not store the contest name (just the column number in the spreadsheet -- an example is below). 

A route that should get 90%+ of the cases for now is for you to read my cleaned/combined file (which matches rows by county + office + district + party + rank within party). `candidate_h` is Snyder's name, `candidate_m` is yours. 

``` r
library(tidyverse)

out <- readxl::read_excel("combined/compare.xlsx", sheet = 1)

out |> 
  filter(!is.na(candidate_m),
         party_detailed != "UNDERVOTE") |> 
  select(state, county_name, office, district, 
         candidate_m,
         candidate_h,
         party_detailed_h = party_detailed)
#> # A tibble: 7,246 × 7
#>    state  county_name office   district candidate_m candidate_h party_detailed_h
#>    <chr>  <chr>       <chr>    <chr>    <chr>       <chr>       <chr>           
#>  1 ALASKA STATEWIDE   US SENA… ALASKA   GROSS AL    GROSS, AL   DEMOCRAT        
#>  2 ALASKA STATEWIDE   US SENA… ALASKA   SULLIVAN D… SULLIVAN, … REPUBLICAN      
#>  3 ALASKA STATEWIDE   US PRES… FEDERAL  JOSEPH R B… JOSEPH R. … DEMOCRAT        
#>  4 ALASKA STATEWIDE   US PRES… FEDERAL  JAMES G JA… JANOS, JAM… GREEN           
#>  5 ALASKA STATEWIDE   US PRES… FEDERAL  JO JORGENS… JO JORGENS… LIBERTARIAN     
#>  6 ALASKA STATEWIDE   US PRES… FEDERAL  DONALD J T… DONALD J. … REPUBLICAN      
#>  7 ALASKA STATEWIDE   US HOUSE 000      GALVIN ALY… GALVIN, AL… DEMOCRAT        
#>  8 ALASKA STATEWIDE   US HOUSE 000      YOUNG DON   YOUNG, DON  REPUBLICAN      
#>  9 ALASKA STATEWIDE   STATE S… 00B      MYERS ROBE… MYERS, ROB… REPUBLICAN      
#> 10 ALASKA STATEWIDE   STATE S… 00D      LAMB THOMAS LAMB, THOM… DEMOCRAT        
#> # ℹ 7,236 more rows
```

<sup>Created on 2024-05-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

This file currently limits to parties c("DEMOCRAT", "REPUBLICAN", "LIBERTARIAN", "GREEN",  "UNDERVOTE",  , "WRITEIN")


Jim's original metadata looks like:

```
Rows: 92,979
Columns: 34
$ state             <chr> "AK", "AK", "AK", "AK", "AK", "AK", "AK", "AK", "AK"…
$ county            <chr> "Statewide", "Statewide", "Statewide", "Statewide", …
$ place             <chr> "", "", "", "", "", "", "", "", "", "", "", "", "", …
$ column            <dbl> 9, 9, 10, 11, 12, 13, 14, 15, 16, 17, 17, 18, 19, 20…
$ item              <chr> "US_PRES", "US_PRES", "US_PRES", "US_PRES", "US_PRES…
$ dist              <chr> "", "", "", "", "", "", "", "", "", "", "", "", "", …
$ yes_orientation   <chr> "", "", "", "", "", "", "", "", "", "", "", "", "", …
$ description       <chr> "", "", "", "", "", "", "", "", "", "", "", "", "", …
$ choice_id         <dbl> 0, 2, 1, 1, 1, 1, 1, 1, 1, 0, 2, 1, 1, 1, 0, 1, 1, 1…
$ choice            <chr> "UNDERVOTE?", "JOSEPH R. BIDEN", "DON BLANKENSHIP", …
$ party             <chr> "", "DEM", "CON", "AIP", "GRN", "LBT", "NOM", "REP",…
$ level             <chr> "N", "N", "N", "N", "N", "N", "N", "N", "N", "N", "N…
$ spending          <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ num_votes         <dbl> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
$ measure           <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0…
$ retention         <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0…
$ nonpartisan       <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0…
$ unexp_term        <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ incumbent         <dbl> 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0…
```

**Comment by kuriwaki** [comment]
@mreece13 take a look at https://github.com/kuriwaki/cvr_harvard-mit_scripts/tree/harv-stata-parquet/R/build-harvard/metadata that includes the metadata based on columns. 

The "column" is the column number of the contest, but its construction differs for JSON files where there is no obvious spreadsheet. For those, the `column` is often the alphabetical order of the contest.

**Comment by mreece13** [comment]
It looks like https://github.com/kuriwaki/cvr_harvard-mit_scripts/tree/harv-stata-parquet/R/build-harvard/metadata/contests_snyder.csv does not include the `column` variable, it might have been dropped accidentally along the way. Similarly, the file does not appear to be particularly clean, so I just want to double-check that this is the correct one. For an example, there are 1190 unique presidential candidates listed in the metadata, which doesn't seem quite right.

**Comment by kuriwaki** [comment]
Sorry I inadvertentlg dropped `column` in the script in fmt_harv_to_medsl()
before uploading. Will upload late tonight.

The proliferation of rows is in large part due to NJ, where there is one
column per every write-in.

**Comment by kuriwaki** [comment]
I do continue to think that reconstructing the linkage by constructing the
cand_rank variable as in the approach of
https://github.com/kuriwaki/cvr_harvard-mit_scripts/blob/main/R/combine/02_combine-to-wide.R

is easier than trying to work with the `column` column

**Comment by mreece13** [comment]
Okay, thanks for the upload. I am actually not sure it matters too much, I have just taken the version of the Harvard data in `CVR_parquet` and used a fuzzy, blocked string join function (written by a Yale pre-doc) to merge together the two versions we have. A couple notes:
- The merge is about 85% accurate. There are also ~100 rows that return duplicate matches, but they are all accurate and represent typos on either the Harvard or MEDSL side that we should fix.
- It's a bit less apparent to me that the metadata of either team is "more" clean than the others. Neither of us clearly standardize a format for names (e.g., we both switch between "last, first" and "first last" all the time) and the main difference appears to be that the MEDSL team strips out all punctuation while Harvard keeps it. I think we should probably run this type of merge and then assign some people to go through and standardize our names manually based on an agreed format. This would be an additional "manual" step, but it would increase the quality a lot, and is relatively easy to add as a step in the pipeline. 

Below is the code I used for identifying candidates and then merging. 

```r
library(zoomerjoin)
library(arrow)
library(tidyverse)

contests_harvard = open_dataset("../CVR_parquet/harvard/") |> 
  distinct(state, office, district, candidate) |>
  collect()

contests_medsl = open_dataset("../CVR_parquet/medsl/") |> 
  distinct(state, office, district, candidate) |>
  collect()

contests_match = jaccard_left_join(
  contests_medsl, contests_harvard, 
  n_gram_width=6, n_bands=50, band_width=6,
  by = "candidate", 
  clean = TRUE,
  block_by = c("state", "office", "district")
)

# show the number of duplicates
contests_duplicates = contests_match |> 
  filter(n() > 1, .by = c("state.x", "office.x", "district.x", "candidate.x")) |> 
  arrange(state.x, office.x, district.x, candidate.x) |> 
  # this removes a lot of the UNDERVOTE? vs UNDERVOTE in Snyder
  filter(candidate.x != "UNDERVOTE")
```

**Comment by kuriwaki** [comment]
@mreece13, I had sensed initially that MEDSL did less standardization of a given person's name within a state (e.g. a U.S. Senate candidate's name is formatted differently in different counties). But I can't find examples of that issue now, so I agree that the name standardization in the two teams look similar. 

In that case, I would say the name standardization is not really a priority for release. There is an argument for leaving the names as they were printed on the ballot. And I think standardizing last vs. first name order is pretty difficult in the presence of various formats. But if it is fast, sure. 

(Ben Green, author of zoomerjoin, is coming to Yale for his PhD in polisci + stats after his predoc!)

**Comment by kuriwaki** [comment]
By the way, this stringdist join was helpful for identifying the remaining cases where party did not match but the name does. I analyzed them and gave them all issues (#118, #117, #116, #115, #114)

**Comment by mreece13** [comment]
Okay, that makes sense. It was also my initial sense that we had done less standardization of the names but that appears to be less of the case. Either way, I'm glad it helped identify some additional edge cases that we can resolve. We can work on name standardization later, I also agree that leaving them as is on the ballot has some advantages as well. 

(I also met Ben at MIT's admitted students day and attempted to convince him to join me)
